Fix: Run doctests for structs with lifetime parameters from IDE

anergictcell opened the pull request:
Fixes #14142: Doctests can't be triggered for structs with lifetimes

This MR adds lifetime parameters to the structs path for runnables so that they can be triggered from an IDE as well.

This is my first MR for rust-analyzer, please let me know if I should change something, either in code or the description here.

HKalbasi commented:
@anergictcell Are you sure your problem is fixed with this PR? I don't think cargo needs lifetime names for tests, as I can run your examples in #14142 in vscode without any problem.

anergictcell commented:
Hi @HKalbasi 

Weird, it does not work for me in either SublimeText or VSCode. I made a screen capture and uploaded it below.
It does not show an error, but it simply ignores the test. 

VScode:
`rust-analyzer version: 0.0.0 (a6603fc21 2023-02-16)`

SublimeText:
`rust-analyzer 0.0.0 (646f97385 2023-02-12)`

Rust version:
`rustup 1.24.3 (2021-05-31)
info: This is the version for the rustup toolchain manager, not the rustc compiler.
info: The currently active rustc version is rustc 1.66.1 (90743e729 2023-01-10)`

OS:
`MacOS 13.2 M1`

https://user-images.githubusercontent.com/875703/221000912-f09f3f5e-4e07-42e1-8acc-1b874ac25a27.mov

HKalbasi commented:
Ah, my bad. I saw the `ok` in running 0 tests as working.

The exact same problem exists for const generics. Can you address that in addition to lifetimes?

anergictcell commented:
@HKalbasi : I Added the const generic parameters as well, including 2 unit tests. I'm not 100% happy with my solution, because I had to allocate `Vec` to satisfy the borrow checker. Can you have a look and maybe help me with improving - if necessary?
Thanks

HKalbasi commented:
I feel there are still two cases that it can go wrong. One is things with mixed type and const generics, like:
```Rust
struct Foo<T, const N: usize, U, const M: usize>(T, U);
```
I think your implementation will put constants at the end (`Foo<T, U, N, M>` instead of `Foo<T, N, U, M>`), and I'm not sure if `cargo doc` will accept that or not. Iterating over types and consts at the same time might also remove the need for vec allocation, but I don't think it would be deal breaker if it doesn't and we need a vec allocation anyway.

Another thing which I'm not sure about is impl with concrete constant:
```Rust
struct Foo<const N: usize>();

impl Foo<2> {
  /// doc test
}

impl Foo<5> {
  /// doc test
}
```
I think your implementation will pick `Foo<N>` and it doesn't work, but I might be wrong. A test for this case would be good.

anergictcell commented:
Thanks a lot for the fast response.
You're right, my implementation will not work for the two cases you're listing. I misunderstood the [docs](https://doc.rust-lang.org/reference/items/generics.html) that the order will always be Lifetimes -> Types -> Consts - but that is apparently wrong and the order is not fixed.
This means I will have to find another approach to extract the info about types and consts. I saw a `hir::Type::generic_params` method, I'll check if this can be used to get ordered parameters and then try to work my way from there. Otherwise I will have to dig deeper. This is my first dive into rust-analyzer and I was quite happy I figured out how to get the lifetime info and then the consts as well.

HKalbasi commented:
Lifetimes are always first but types and consts can be mixed. You can iterate over them in order with `.type_or_consts` method iirc.

Another approach could be finding the impl, and copy the type of it from user code verbatimly. I guess it will always work.

If you found that you can't put more work on this, we can merge it and defer the rest to future PRs. This PR is a clear improvement even at the current state.

anergictcell commented:
I'll give it a try and see if I can fix it. It's an interesting challenge and learning experience. If it gets too tricky, I'll let you know and we can merge the current version.

anergictcell commented:
I managed to also include `const` in the doctest runnables now. I refactored my code a bit, which also makes it cleaner and nicer and added more tests as well.

bors commented:
:umbrella: The latest upstream changes (presumably #14216) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

anergictcell commented:
I fixed the merge conflicts that bors was complaining about. Is there anything I need to do to indicate this?

HKalbasi commented:
I don't have r+ access but this looks good to me now.

r? @Veykril

Veykril commented on crates/hir/src/lib.rs:
```diff
@@ -3339,6 +3376,58 @@ impl Type {
+    }
+
+    /// Combines lifetime indicators, type and constant parameters into a single `Iterator`
+    pub fn lifetime_type_const_paramaters<'a>(
```
```suggestion
    pub fn generic_parameters<'a>(
```

anergictcell replied:
That makes sense, thanks. I changed it with a new commit.

Veykril commented:
You should have bors permissions now as well

HKalbasi commented:
@bors r+

bors commented:
:pushpin: Commit af79491ae62ebba739d3375c425b71c48f4b4905 has been approved by `HKalbasi`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust-analyzer) for this repository.

<!-- @bors r=HKalbasi af79491ae62ebba739d3375c425b71c48f4b4905 -->
<!-- homu: {"type":"Approved","sha":"af79491ae62ebba739d3375c425b71c48f4b4905","approver":"HKalbasi","queue":"https://bors.rust-lang.org/homu/queue/rust-analyzer"} -->

bors commented:
:hourglass: Testing commit af79491ae62ebba739d3375c425b71c48f4b4905 with merge c386316fe0aae4f49a76dec8c81c0bd560c50e77...
<!-- homu: {"type":"BuildStarted","head_sha":"af79491ae62ebba739d3375c425b71c48f4b4905","merge_sha":"c386316fe0aae4f49a76dec8c81c0bd560c50e77"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust-analyzer/actions/runs/4291651628/jobs/7477358235)
Approved by: HKalbasi
Pushing c386316fe0aae4f49a76dec8c81c0bd560c50e77 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"HKalbasi","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/rust-analyzer/actions/runs/4291651628/jobs/7477358235"},"merge_sha":"c386316fe0aae4f49a76dec8c81c0bd560c50e77"} -->
